### PR TITLE
Add PMSN model

### PR DESCRIPTION
**Pull request by guarin**

### CODE REVIEW

File: examples/pytorch/msn.py [hunk 1]
```diff
@@ -21,7 +21,7 @@ def __init__(self, vit):
 
         self.mask_ratio = 0.15
         self.backbone = MAEBackbone.from_vit(vit)
-        self.projection_head = MSNProjectionHead()
+        self.projection_head = MSNProjectionHead(384)
```
Comment (guarin) [inline examples/pytorch/msn.py:24]: Unrelated, noticed that the correct dimension was missing and fixed it.

File: examples/pytorch/msn.py [hunk 2]
```diff
@@ -84,6 +84,9 @@ def forward_masked(self, images):
 )
 
 criterion = MSNLoss()
+# Or for PMSN:
```
Comment (IgorSusmelj) [inline examples/pytorch/msn.py:87]: I'd create separate examples as we did before. If someone is interested in PMSN they can find it under examples and use it right away. It also shows exactly that everything except the loss is the same. We did the same for NNCLR and other models.

**Comment by guarin** [comment]
Added example code and docs for PMSN
[PMSN — lightly 1.3.1 documentation.pdf](https://github.com/lightly-ai/lightly/files/10969890/PMSN.lightly.1.3.1.documentation.pdf)
[MSN — lightly 1.3.1 documentation.pdf](https://github.com/lightly-ai/lightly/files/10969891/MSN.lightly.1.3.1.documentation.pdf)

File: docs/source/getting_started/benchmarks/cifar10_benchmark.py [hunk 3]
```diff
@@ -127,7 +127,7 @@
     # reduce batch size for distributed training
     batch_size = batch_size // devices
 else:
-    strategy = "auto"  # Set to None if using PyTorch Lightning < 2.0
+    strategy = None  # Set to "auto" if using PyTorch Lightning >= 2.0
     # limit to single device if not using distributed training
```
Comment (guarin) [inline docs/source/getting_started/benchmarks/cifar10_benchmark.py:131]: Had to change this back to PyTorch Lightning < 2.0 because other changes are required to make the benchmarks fully compatible with version >= 2.0. See #1112

**Review by MalteEbner — COMMENTED** [review]
I did not find any problems, looks good.

Could you describe how you came up with the four new files? E.g. tell from where they were copied and add a comment what changed after copying.

File: docs/source/getting_started/benchmarks/imagenette_benchmark.py [hunk 4]
```diff
@@ -877,6 +879,83 @@ def configure_optimizers(self):
...
+        self.anchor_backbone = copy.deepcopy(self.backbone)
+        self.anchor_projection_head = copy.deepcopy(self.projection_head)
+
+        utils.deactivate_requires_grad(self.backbone)
+        utils.deactivate_requires_grad(self.projection_head)
+
+        self.prototypes = nn.Linear(256, 1024, bias=False).weight
+        self.criterion = MSNLoss(
+            target_distribution="power_law", power_law_exponent=0.25
+        )
```
[the remark this answers is not in the document]
Comment (guarin) [reply docs/source/getting_started/benchmarks/imagenette_benchmark.py:908]: Yes!

File: tests/loss/test_MSNLoss.py [hunk 5]
```diff
@@ -38,76 +71,127 @@ def test_sharpen(self, seed=0):
...
-        criterion = MSNLoss()
-        anchors = torch.rand((8 * 2, 10)).cuda()
-        targets = torch.rand((8, 10)).cuda()
-        prototypes = torch.rand((4, 10), requires_grad=True).cuda()
-        criterion(anchors, targets, prototypes)
+        for target_distribution in [
+            "uniform",
+            "power_law",
+            _linear_target_distribution,
+        ]:
```
[the remark this answers is not in the document]
Comment (guarin) [reply tests/loss/test_MSNLoss.py:110]: Cleaning up the tests is already planned. I tried to make minimal changes in this PR.

**Review by MalteEbner — COMMENTED** [review]
The structure of the MSN loss is much better now, but I still see potential.

File: lightly/loss/msn_loss.py [hunk 6]
```diff
@@ -200,13 +273,39 @@ def forward(
...
+                )
+                reg_loss = F.kl_div(
+                    input=mean_anchor_probs, target=power_dist, reduction="sum"
+                )
+            else:
+                assert isinstance(self.target_distribution, Callable)
+                target_dist = self.target_distribution(mean_anchor_probs)
+                reg_loss = F.kl_div(
+                    input=mean_anchor_probs, target=target_dist, reduction="sum"
+                )
```
[the remark this answers is not in the document]
Comment (guarin) [reply lightly/loss/msn_loss.py]: I added a `regularization_loss` method to the class and then overwrite the methods in `PMSNLoss` and `PMSNCustomLoss`. This should make it simpler and we don't have to deal with incompatible parameter combinations.

I also didn't make `PMSNLoss` inherit from `PMSNCustomLoss` as the code is straight forward and the extra inheritance makes it harder to understand in my eyes.

**Review by MalteEbner — APPROVED** [review]
Good design! Great that we don't need those two extra attributes.
